### PR TITLE
Migrate kl_div_backward from CUDA_tensor_apply3 to TensorIterator

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34026 Completely kill CUDA_tensor_apply3
* #34025 Migrate lerp from CUDA_tensor_apply3 to TensorIterator
* #34024 Migrate dropout inference from CUDA_tensor_apply3 to TensorIterator
* #34023 Migrate bce loss from CUDA_tensor_apply3 to TensorIterator
* **#34022 Migrate kl_div_backward from CUDA_tensor_apply3 to TensorIterator**
* #34021 Migrate dirichlet from CUDA_tensor_apply3 to TensorIterator
* #34020 Migrate gamma grad from CUDA_tensor_apply3 to TensorIterator

Differential Revision: [D20196080](https://our.internmc.facebook.com/intern/diff/D20196080)

### CODE REVIEW

File: aten/src/ATen/native/cuda/Loss.cu [hunk 1]
```diff
@@ -82,14 +70,19 @@ void binary_cross_entropy_backward_out_kernel(Tensor& grad_input, const Tensor&
...
-  Tensor grad_expand = grad.expand_as(input);
+  auto grad_input = at::empty_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  TensorIterator iter;
+  iter.add_output(grad_input);
+  iter.add_input(target);
+  iter.add_input(grad);
+  iter.build();
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "kl_div_backward_cuda", [&]() {
-    kl_div_backward_kernel<scalar_t>(grad_input, target, grad_expand);
+    scalar_t inv = (reduction == at::Reduction::Mean) ? scalar_t(1.0 / input.numel()) : scalar_t(1.0);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cuda/Loss.cu]: This might introduce slowdown due to additional operation. Can you please add benchmarks to the PR.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Loss.cu]: It is actually not a slowdown, but I fused the
```C++
  if (reduction == at::Reduction::Mean) {	
    return grad_input / input.numel();	
  }
```
into the iterator, so it should be faster. But I will do a benchmark.

File: aten/src/ATen/native/cuda/Loss.cu [hunk 2]
```diff
@@ -82,14 +70,19 @@ void binary_cross_entropy_backward_out_kernel(Tensor& grad_input, const Tensor&
 namespace at { namespace native {
 
 Tensor kl_div_backward_cuda(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction) {
-  auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  Tensor grad_expand = grad.expand_as(input);
+  auto grad_input = at::empty_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cuda/Loss.cu]: You can drop `LEGACY_CONTIGUOUS_MEMORY_FORMAT` here as TensorIterator can work with non-contiguous tensors.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Loss.cu]: fixed

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin On small tensors, the performance is improved. On large tensors, the performance are the same.

```python
import torch
print(torch.__version__)

# warm up
for i in range(1000):
    torch.arange(100000, device='cuda')
    
def benchmark(m, size):
    s1 = 2 ** (size // 2)
    s2 = 2 ** (size - size // 2)
    a = torch.zeros((s1, s2), device='cuda')
    b = torch.ones_like(a, requires_grad=True)
    loss = m(a, b)
    torch.cuda.synchronize()
    %timeit loss.backward(retain_graph=True); torch.cuda.synchronize()

print("reduction = mean:")
m = torch.nn.KLDivLoss().cuda()
for i in range(15, 28):
    benchmark(m, i)

print()
print("reduction = sum:")
m2 = torch.nn.KLDivLoss(reduction="sum").cuda()
for i in range(15, 28):
    benchmark(m2, i)
```

Before:
```
1.5.0a0+9956a23
reduction = mean:
179 µs ± 2.05 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
177 µs ± 2.99 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
173 µs ± 568 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
215 µs ± 4.64 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
376 µs ± 3.43 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
678 µs ± 478 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.29 ms ± 650 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.55 ms ± 15.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.01 ms ± 16.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.96 ms ± 733 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
19.8 ms ± 990 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
39.4 ms ± 56.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
78.7 ms ± 43.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

reduction = sum:
164 µs ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
159 µs ± 5.19 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
154 µs ± 150 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
209 µs ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
332 µs ± 2.29 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
592 µs ± 1.25 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.13 ms ± 337 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.21 ms ± 5.12 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.37 ms ± 13.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
8.72 ms ± 3.71 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
17.3 ms ± 5.41 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
34.5 ms ± 49.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
68.9 ms ± 52.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

After:
```
1.5.0a0+fdac34c
reduction = mean:
113 µs ± 32.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
159 µs ± 2.79 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
162 µs ± 34.2 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
215 µs ± 619 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
372 µs ± 1.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
664 µs ± 438 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.28 ms ± 599 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.51 ms ± 4.95 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.99 ms ± 5.42 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.95 ms ± 554 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
19.8 ms ± 866 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
39.4 ms ± 56.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
78.7 ms ± 46.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

reduction = sum:
142 µs ± 336 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
142 µs ± 222 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
143 µs ± 1.74 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
191 µs ± 1.31 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
328 µs ± 575 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
585 µs ± 523 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.12 ms ± 384 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.2 ms ± 5.02 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.38 ms ± 5.69 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
8.72 ms ± 808 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
17.3 ms ± 4.93 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
34.5 ms ± 49.3 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
68.9 ms ± 51.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

File: aten/src/ATen/native/cuda/Loss.cu (hunk 1 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/cuda/Loss.cu]: I recognize the intend, but fusion requires two kernels for  `reduction == at::Reduction::Mean` and without it. Your version _might_ force additional multiplication call (depends from compiler optimization).

**Review by VitalyFedyunin — APPROVED** [review]
